src/main/scala/combinator/BanditSelector.scala

```diff
@@ -0,0 +1,449 @@
+      case AfterEveryMove =>
+        updateWeight(neighborhood)
+        clearStats(neighborhood)
+      case afterNMoves @ AfterNMoves(_) =>
+        afterNMoves.incrementCounter()
+        if (afterNMoves.isCriterionMet) {
+          updateWeights()
+          afterNMoves.resetCounter()
+          clearStats(neighborhood)
+        }
+      case AfterEveryDescent =>
+    }
+  }
+
+  /** Add a newly collected statistic
+    *
+    * @param neighborhoodStats
+    *   statistics about a move
+    * @param neighborhood
+    *   neighborhood having performed the move
+    */
+  protected def appendStats(
+                             neighborhoodStats: NeighborhoodStats,
+                             neighborhood: Neighborhood
+  ): Unit = {
+    maxSlope = Math.max(maxSlope, neighborhoodStats.slope)
+    maxRunTimeNano = Math.max(maxRunTimeNano, neighborhoodStats.timeNano)
+    val idx = neighborhoodIdx(neighborhood)
+    stats(idx).append(neighborhoodStats)
+  }
+
+  object authorizedNeighborhoodIterator {
+    def apply(): Iterator[Int] = {
+      authorizedNeighborhood.indices.iterator.filter(authorizedNeighborhood(_))
+    }
+  }
+
+  /** Mark a neighborhood as tabu This has the side-effect of decreasing the sum of valid weights
+    *
+    * @param neighborhood
+    *   neighborhood to mark as tabu
+    */
+  def setTabu(neighborhood: Neighborhood): Unit = {
+    val idx = neighborhoodIdx(neighborhood)
+    authorizedNeighborhood(idx) = false
+    sumWeightsValidNeighborhoods -= weights(idx)
+    nTabu += 1
+  }
+
+  def isTabu(neighborhood: Neighborhood): Boolean = {
+    val idx = neighborhoodIdx(neighborhood)
+    authorizedNeighborhood(idx)
+  }
+
+  def isTabu(idx: Int): Boolean = {
+    authorizedNeighborhood(idx)
+  }
+
+  /** Reset the list of tabu neighborhoods and update the sum of weights for valid neighborhoods
+    */
+  private def resetTabu(): Unit = {
+    if (nTabu != 0) {
+      for (idx <- authorizedNeighborhood.indices) {
+        authorizedNeighborhood(idx) = true
+        sumWeightsValidNeighborhoods += weights(idx)
+      }
+    }
+    nTabu = 0
+  }
+
+  /** Tells the probability of choosing a neighborhood, proportional to its weight.
+    *
+    * A neighborhood set as tabu has a probability equal to 0.
+    *
+    * @param neighborhood
+    *   the considered neighborhood
+    */
+  def probability(neighborhood: Neighborhood): Double = {
+    val idx = neighborhoodIdx(neighborhood)
+    if (authorizedNeighborhood(idx)) {
+      weights(idx) / sumWeightsValidNeighborhoods
+    } else {
+      0
+    }
+  }
+
+  /** Tells the weight associated to a neighborhood. There are no bounds set on the weight: it can
+    * be larger than 1.
+    *
+    * @param neighborhood
+    *   the considered neighborhood
+    */
+  def weight(neighborhood: Neighborhood): Double = {
+    val idx = neighborhoodIdx(neighborhood)
+    this.weights(idx)
+  }
+
+  /** Returns the neighborhood with the best weight, if applicable.
```

Comment:
maybe change the "if applicable" with "if non tabu neighborhood exist" ?
```suggestion
  /** Returns the neighborhood with the best weight, if on tabu neighborhood exist.
```

Reply:
@FabianGermeau you mean the following, right?
```suggestion
  /** Returns the neighborhood with the best weight, if no tabu neighborhood exists.
```

Reply:
yes ^^